### PR TITLE
Try to reactivate Multi-J RZ PSATD checksums

### PR DESCRIPTION
**Pull request by NeilZaim**
The checksums should fail for now because they haven't been reset since #2377 but I want to see if I get the same checksums on Azure and locally (cf. #2132).

### CODE REVIEW

**Comment by EZoni** [comment]
Just for comparison, I tried to run the test locally on this branch and got these benchmark changes:
```
./analysis_default_regression.py  multi_J_rz_psatd_plt00050
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bx]
Benchmark: [lev=0,Bx] 7.576533012513204e-12
Plotfile : [lev=0,Bx] 8.052209202211183e-12
Absolute error: 4.76e-13
Relative error: 6.28e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bz]
Benchmark: [lev=0,Bz] 4.005984970626178e-12
Plotfile : [lev=0,Bz] 4.929178810213553e-12
Absolute error: 9.23e-13
Relative error: 2.30e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ey]
Benchmark: [lev=0,Ey] 2.123908815458177e-03
Plotfile : [lev=0,Ey] 2.606440568261221e-03
Absolute error: 4.83e-04
Relative error: 2.27e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,jy]
Benchmark: [lev=0,jy] 6.317201519130083e-02
Plotfile : [lev=0,jy] 5.629517976515543e-02
Absolute error: 6.88e-03
Relative error: 1.09e-01
```

Azure seems to report these ones, instead:
```
./analysis_default_regression.py  multi_J_rz_psatd_plt00050
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bx]
Benchmark: [lev=0,Bx] 7.576533012513204e-12
Plotfile : [lev=0,Bx] 6.631769818131767e-12
Absolute error: 9.45e-13
Relative error: 1.25e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bz]
Benchmark: [lev=0,Bz] 4.005984970626178e-12
Plotfile : [lev=0,Bz] 3.591943859588349e-12
Absolute error: 4.14e-13
Relative error: 1.03e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ey]
Benchmark: [lev=0,Ey] 2.123908815458177e-03
Plotfile : [lev=0,Ey] 1.818107609464795e-03
Absolute error: 3.06e-04
Relative error: 1.44e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,jy]
Benchmark: [lev=0,jy] 6.317201519130083e-02
Plotfile : [lev=0,jy] 4.261308266365557e-02
Absolute error: 2.06e-02
Relative error: 3.25e-01
```
I'm curious to know if you get the Azure ones locally, or another additional set of values.

**Comment by NeilZaim** [comment]
So I tried locally with 3 different versions of blas/lapack and I got 3 different results... each different than what you got and than what Azure got :laughing: . I don't know if this means that the fluctuations are just due to differing versions of blas/lapack (which I reckon is possible) or if the fluctuations are unrelated (or maybe both :fearful: ).

@EZoni What version of blas/lapack did you use for your local test?

I copy-paste below the different outputs I got.
**Blas/Lapack installed with conda using the conda-forge channel (liblapack and libblas 3.9.0, blaspp and lapackpp 2020.10.02)**
```
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bx]
Benchmark: [lev=0,Bx] 7.576533012513204e-12
Plotfile : [lev=0,Bx] 7.620588373653700e-12
Absolute error: 4.41e-14
Relative error: 5.81e-03
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bz]
Benchmark: [lev=0,Bz] 4.005984970626178e-12
Plotfile : [lev=0,Bz] 4.312150974887443e-12
Absolute error: 3.06e-13
Relative error: 7.64e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ey]
Benchmark: [lev=0,Ey] 2.123908815458177e-03
Plotfile : [lev=0,Ey] 2.073450921207310e-03
Absolute error: 5.05e-05
Relative error: 2.38e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,jy]
Benchmark: [lev=0,jy] 6.317201519130083e-02
Plotfile : [lev=0,jy] 4.236822131209150e-02
Absolute error: 2.08e-02
Relative error: 3.29e-01
```

**Still with conda but libblas and liblapack downgraded to 3.8.0 using the nogil channel (blaspp and lapackpp packages unchanged compared to previous case)**

```
Benchmark: [lev=0,Bx] 7.576533012513204e-12
Plotfile : [lev=0,Bx] 7.357295158251789e-12
Absolute error: 2.19e-13
Relative error: 2.89e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bz]
Benchmark: [lev=0,Bz] 4.005984970626178e-12
Plotfile : [lev=0,Bz] 4.129644332371564e-12
Absolute error: 1.24e-13
Relative error: 3.09e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ey]
Benchmark: [lev=0,Ey] 2.123908815458177e-03
Plotfile : [lev=0,Ey] 2.162655037068320e-03
Absolute error: 3.87e-05
Relative error: 1.82e-02
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,jy]
Benchmark: [lev=0,jy] 6.317201519130083e-02
Plotfile : [lev=0,jy] 5.006425455141275e-02
Absolute error: 1.31e-02
Relative error: 2.07e-01
```

**Blas/Lapack installed with apt with Debian (libblas-dev and liblapack-dev 3.9.0-3) and blaspp and lapackpp installed with cmake using the git repo (same as in the Azure CI tests)**

```
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bx]
Benchmark: [lev=0,Bx] 7.576533012513204e-12
Plotfile : [lev=0,Bx] 6.755086400678032e-12
Absolute error: 8.21e-13
Relative error: 1.08e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Bz]
Benchmark: [lev=0,Bz] 4.005984970626178e-12
Plotfile : [lev=0,Bz] 3.274337266080892e-12
Absolute error: 7.32e-13
Relative error: 1.83e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,Ey]
Benchmark: [lev=0,Ey] 2.123908815458177e-03
Plotfile : [lev=0,Ey] 1.871795314999358e-03
Absolute error: 2.52e-04
Relative error: 1.19e-01
ERROR: Benchmark and plotfile checksum have different value for key [lev=0,jy]
Benchmark: [lev=0,jy] 6.317201519130083e-02
Plotfile : [lev=0,jy] 4.119572827904305e-02
Absolute error: 2.20e-02
```

The installation of blas/lapack in the last case is very similar to what we have with Azure, except that Azure uses Ubuntu 20.04 which has libblas-dev and liblapack-dev 3.9.0-1 and I have version 3.9.0-3 on my Debian machine. I don't know if this is enough to make a difference.

**Comment by NeilZaim** [comment]
I've tried to reset the benchmarks after #2523. Although this fix affects the checksum values, I still see discrepancies between the values I see in Azure and the values I have locally.

**Comment by NeilZaim** [comment]
Tried locally with the same blas and lapack version as in Azure (libblas-dev and liblapack-dev 3.9.0-1 installed with apt from ubuntu 20.04) and I still get different results than with Azure. So the issue does not seem to be linked with the blas and lapack versions.

**Comment by NeilZaim** [comment]
So I've looked a bit more into why the checksum values are fluctuating for this test. It looks like it is actually coming from the lapack calls.

More specifically, I see that the result of this call (in the initialization of the matrix used for the Hankel transform) can vary to machine precision.

https://github.com/ECP-WarpX/WarpX/blob/cd3488e0b0547ac9b7fb4d68577ab1269c598fd2/Source/FieldSolver/SpectralSolver/SpectralHankelTransform/HankelTransform.cpp#L170

- On my laptop, I've checked that the resulting `M` matrix is different for 2 different versions of lapack/blas, although all the inputs are exactly the same.
- I also get a slightly different set of values in Azure for the `M` matrix, even though I thought it was using the same version of blas/lapack as in one of the 2 cases on my laptop (I find this a bit surprising, but maybe this is expected, I don't know exactly what is done in this lapack call).

So assuming that this is correct, the question is now what to do with this benchmark:
- Reactivate it and keep in mind that the values obtained locally will generally be different (maybe a bit annoying)?
- Reactivate it but increase tolerance and/or remove values that are too sensitive to machine precision errors?
- Don't use benchmarks for RZ-PSATD tests (maybe a bit dangerous)?
- Try to find a way to make the calls give the same results to machine precision (I don't know if that's doable)?

Any opinion on this @EZoni @RemiLehe @ax3l ?

**Comment by EZoni** [comment]
Thank you for doing all these tests, @NeilZaim. Among the possible options that you mentioned, the following one would seem to me like the best way to go:

> Reactivate it but increase tolerance and/or remove values that are too sensitive to machine precision errors?

**Comment by ax3l** [comment]
What's the physics in `Bx`, `Bz` and `Ey` and `jy` for these RZ tests? cc @RemiLehe 

Should they be close to zero? Our checksum defaults don't work well for fields close to zero unless you adjust the super-low default `atol` for the `check()` accordingly in a custom python analysis script:
https://github.com/ECP-WarpX/WarpX/blob/3a0446c3897a94e804b6a531150f0388a0e813c0/Regression/Checksum/checksumAPI.py#L38-L51

Let's take from
https://github.com/ECP-WarpX/WarpX/blob/3a0446c3897a94e804b6a531150f0388a0e813c0/Examples/Tests/PML/analysis_pml_psatd_rz.py#L55
and change the tolerance or ship these close-to-zero fields in output as I did in #2909

**Comment by ax3l** [comment]
btw, that `atol` does not work for single precision, which ranges only in the `1.2e-38` to `3.4e+38` range...
(the default of 9 orders of magnitude for `rtol` also does not work for single precision.)

File: Regression/WarpX-tests.ini [hunk 1]
```diff
@@ -2400,16 +2400,15 @@ dim = 2
...
+numprocs = 1
+useOMP = 0
 numthreads = 1
 compileTest = 0
 doVis = 0
 compareParticles = 1
 particleTypes = driver plasma_e plasma_p
-# TODO: intentionally empty because still unstable
-analysisRoutine =
+analysisRoutine = Examples/analysis_default_regression.py
```
Comment (ax3l) [inline Regression/WarpX-tests.ini:2538]: Probably solution: skip some close-to-zero fields in output OR use a custom analysis to change tolerance, see comments and regular `PSATD + RZ` tests in, e.g., PML

**Comment by NeilZaim** [comment]
So in the end I've done as you suggested @ax3l and I've removed the close-to-zero fields from the outputs.

I've also had to reset the benchmarks but that's not surprising since this PR is from ~7 months ago.

File: Regression/Checksum/benchmarks_json/multi_J_rz_psatd.json [hunk 2]
```diff
@@ -1,49 +1,45 @@
...
-    "jz": 1930576792579724.0,
-    "rho": 5286010.8543447815,
-    "rho_driver": 6281777.267757398,
-    "rho_plasma_e": 49565978.70143692,
-    "rho_plasma_p": 50769177.377112545
+    "By": 24912.66213622398,
+    "Ex": 4667305981763.072,
+    "Ez": 4307437907855.343,
+    "jx": 362736023475894.4,
+    "jz": 1937267309900551.0,
```
Comment (EZoni) [inline Regression/Checksum/benchmarks_json/multi_J_rz_psatd.json:18]: Thanks for looking again into this, @NeilZaim! I see that you removed the checksums for `Bx` and `Bz` which are indeed close to zero (order `1e-12`), but I see that you also removed the ones for `Ey` and `jy` which are larger instead. I understand that physical units play a role here, but can you confirm that indeed the results were unstable also for `Ey` and `jy` (in other words, confirm that these were not removed by mistake)?

Comment (NeilZaim) [reply Regression/Checksum/benchmarks_json/multi_J_rz_psatd.json:18]: Yes I confirm that `Ey` and `jy` are also unstable.

The value of `Ey` is the same order of magnitude as that of `Bx`and `Bz` (multiplied by `c`) and the value of `jy` is approximately 1e-15 that of `jx` & `jz` so also pretty much zero to machine precision.

**Review by EZoni — APPROVED** [review]
Thank you for looking into this, @NeilZaim! Looks good to me.

**Pull request merged**